cheatsheets/Forgot_Password_Cheat_Sheet.md

```diff
@@ -1,60 +1,137 @@
+> Security considerations for properly handling tokens and codes on the backend are discussed in the [methods' implementation](#methods-implementation) section.
 
-If all submitted data is correct, Step 2 should display at least two of the user's pre-established personal security questions, along with input fields for the answers. It's important that the answer fields are part of a single HTML form.
+### Forgot Password Request
 
-Do not provide a drop-down list for the user to select the questions he wants to answer. Avoid sending the username as a parameter (hidden or otherwise) when the form on this page is submitted. The username should be stored in the server-side session where it can be retrieved as needed.
+The user uses the forgot password service and inputs their username or email. To ensure the security on this stage, the below should implemented:
 
-Because users' security questions / answers generally contains much less entropy than a well-chosen password (how many likely answers are there to the typical *What's your favorite sports team?* or *In what city where you born?* security questions anyway?), make sure you limit the number of guesses attempted and if some threshold is exceeded for that user (say 3 to 5), lock out the user's account for some reasonable duration (say at least 5 minutes) and then challenge the user with some form of challenge token per standard multi-factor workflow (see \#3, below) to mitigate attempts by hackers to guess the questions and reset the user's password. It is not unreasonable to think that a user's email account may have already been compromised, so tokens that do not involve email, such as SMS or a mobile soft-token, are best.
+- Unified user message, whether the user exists or not.
+- Unified time for the user response. A quick exit method should not be used if a user doesn't exist and an email is not sent.
+- Rate-limiting (*e.g.* Captchas, blocking IPs for a period of time, etc.) to protect the application against brute-force attacks.
+- Employ proper [Input Validation](Input_Validation_Cheat_Sheet.md) on the sent field.
 
-## Step 3) Send a Token Over a Side-Channel
+### Code or Token Activation
 
-After step 2, lock out the user's account immediately. Then SMS or utilize some other multi-factor token challenge with a randomly-generated code having 8 or more characters. 
+After a user uses the service to reset their password, a code or token is provided to the user following one of the [methods](#methods) discussed later on in this document. In order to securely handle that token or code once the user inserts it, the same security practices employed for the [forgot password request](#forgot-password-request) should be used with one additional measure that should be used when sending the code or token in the querystring:
```

Comment:
The second sentence here is very hard to read - I'm not entirely sure what it's saying.